src/Symfony/Component/HttpClient/Response/AmpResponse.php

```diff
@@ -152,7 +133,7 @@ public function __wakeup()
     public function __destruct()
     {
         try {
-            $this->doDestruct();
+            //$this->doDestruct();
```

Comment:
I figured out the blocker: we cannot switch fibers in destructors.
Most tests pass now. The ones that rely on destructors fail of course.

I tried this instead:
```php
            if ($this->initializer && null === $this->info['error']) {
                $clone = clone $this;
                $clone->shouldBuffer = true;
                EventLoop::queue($clone->getHeaders(...));
            }
```

It doesn't work (yet?) but I fear it cannot work because it relies on the event loop having a chance to be resumed after the destructor is called, which might not be always the case.

I might miss something how course.

Reply:
I can make many tests pass by patching php and Amp this way.

I'm going to need your help @kelunik and @trowski to figure out if there is something to improve in either PHP, Amp and/or Symfony.


1. allow switching fibers in destructors:

```diff
--- a/Zend/zend_fibers.c
+++ b/Zend/zend_fibers.c
@@ -391,7 +391,7 @@ ZEND_API void zend_fiber_switch_unblock(void)
 
 ZEND_API bool zend_fiber_switch_blocked(void)
 {
-       return zend_fiber_switch_blocking;
+       return false;//zend_fiber_switch_blocking;
 }
```

2. ignore `Cannot suspend in a force-closed fiber`:

```diff
--- a/src/EventLoop/Internal/AbstractDriver.php
+++ b/src/EventLoop/Internal/AbstractDriver.php
@@ -538,6 +538,7 @@ abstract class AbstractDriver implements Driver
     private function createCallbackFiber(): void
     {
         $this->callbackFiber = new \Fiber(function (): void {
+            try {
             do {
                 $this->invokeMicrotasks();
 
@@ -599,6 +600,8 @@ abstract class AbstractDriver implements Driver
                 /** @noinspection PhpUnhandledExceptionInspection */
                 \Fiber::suspend($this->internalSuspensionMarker);
             } while (true);
+            } catch (\FiberError) {
+            }
         });
     }
```

Reply:
Allowing fiber switches in destructors requires the garbage collector to be safe for re-entry from multiple fibers.

It's also generally unadvisable to switch fibers in a finally block, which is likely what is causing the force-closed fiber issue. This is similar to force-closing a generator, you cannot yield in the finally block.

Perhaps the solution then is to move any I/O out of a finally block?

You can switch fibers in a shutdown function, in case that is of any help.

I took a short look at the code this morning and… wow there's a lot going on here through multiple layers. Wrapping `$this->doDestruct()` into `EventLoop::queue($this->doDestruct(...));` had some promising results, but throwing the exceptions from the destructor the event loop. Perhaps wrap that call with a catch that adds the exceptions to a list and re-throws them at a later time? Can `initialize()` be run immediately upon receiving a response rather than waiting until the destructor to run if it wasn't otherwise run?

Reply:
Calling `initialize()` soon after creating an `AmpResponse` instance seems more promising, but I'm getting exceptions thrown in the wrong places in tests I think. I'm going to bow out here for now, I have a rather busy weekend ahead of me. Hopefully @nicolas-grekas you can run with these ideas and make some progress.

Reply:
Thanks for having a look! Calling initialize means waiting for the headers of the response to come back. This means we cannot call it immediately since this would make everything synchronous.

A common pattern with the concurrency model of HttpClient is to stack requests in an array and let them complete when the array is released:

```php
$pool = [];
foreach ($urls as $url) {
	$pool[] = $client->request('GET', $url);
}
unset($pool); // either explicit or more commonly implicit on exit of a method
```

The only way to make this API work is by doing I/O in destructors I'm afraid.

> Wrapping $this->doDestruct() into EventLoop::queue($this->doDestruct(...)); had some promising results, but throwing the exceptions from the destructor to the event loop handler

That's the closest we can have now indeed. But this requires running inside the event loop, which is not the most common for HttpClient. Alternatively, this could work with a new method on the client to wait for requests in such a pool, but this would break the current API, and the current API has one major desired feature: if ppl forget about checking their responses, they will get an exception if a 4xx/5xx is returned. This also requires doing I/O in destructors, and cannot be solved by any extra method of course.

When not running in an event loop, I thought about pooling destructed responses in a queue to check on every other method call on the client/responses. This would have the quite unpleasant effect or running those destructed requests out of context (think call-stack). This also wouldn't work for the very last set of responses. Yes, shutdown function (or $client->reset() as called by eg Symfony DI) could provide this missing tick, but the DX looks terrible...

An alternative idea could be to run in some sort of fiber-less mode inside destructors - that'd be a sync mode I suppose. Is that something that AMP could be of any help for?

> Allowing fiber switches in destructors requires the garbage collector to be safe for re-entry from multiple fibers

Is it correct or not that calling destructors during collect cycles is the step that requires this reentry, and that calling them on refcount=0 doesn't? If correct, would it be possible to allow switching fibers in refcount=0 situations? This would partially fix the problem for HttpClient since response objects are rarely caught is ref cycles.

Of course, the next question is: would it be possible to make the GC safe for re-entry from multiple fibers? The current situation makes destructors kinda reintroduce the “What color is your function?” problem, which fibers were designed to avoid.

Reply:
> ```php
> $pool = [];
> foreach ($urls as $url) {
> 	$pool[] = $client->request('GET', $url);
> }
> unset($pool); // either explicit or more commonly implicit on exit of a method
> ```

That's quite a strange design choice and pattern IMO, which is indeed not possible to implement right now with fibers. In general, destructors shouldn't really do any work, similar to constructors.

> The current situation makes destructors kinda reintroduce the “What color is your function?” problem, which fibers were designed to avoid.

Indeed, but it's avoidable if destructors don't do any work, which worked quite well for us so far.

Reply:
Please open an issue in php-src. I'd like to see support for suspensions in destructors as well. It's been on my (long) todo list, but now might be a good time to move it toward the top. Unfortunately I know little about the GC, so will probably need some help.

We've avoided suspending in destructors by using a defer in the event loop, but this probably only really works in a long running app (and runs the risk of never running). Closing a socket doesn't require suspending because there's no result to await, but it is I/O, so that's a fair criticism.

I glanced through the GC code today, and it looks to already guard against re-entry. The concern of course would be that if a destructor which suspended never again resumed it would block the GC from ever being run again. I'm not sure if this is acceptable or not.